src/middlewared/middlewared/plugins/rsync.py

```diff
@@ -294,7 +294,13 @@ async def validate_rsync_task(self, data, schema):
 
                             # If for whatever reason the dir does not exist, let's create it
                             # An example of this is when we run rsync tests we nuke the directory
-                            await self.middleware.run_in_thread(ssh_dir_path.mkdir, 0o700, False, True)
+                            if not await self.middleware.run_in_thread(ssh_dir_path.exists):
```

Comment:
We're creating 5 separate threads to perform operations that can be done in a single thread. Also, this is a race condition to check for existence before creating.

```
def handle_ssh_dir(self, ssh_dir_path, known_hosts_path, known_hosts_text, uid, gid):
    try:
        ssh_dir_path.mkdir(0o700)
        os.chown(ssh_dir_path, uid, gid)
    except FileExistsError:
        pass
  
    known_hosts_path.write_text(known_hosts_text)
    os.chown(known_hosts_path, uid, gid)

await self.middleware.run_in_thread(
    self.handle_ssh_dir, ssh_dir_path, known_hosts_path, known_hosts_text, user['pw_uid'], user['pw_gid']
)